File: src/Forge.Host/Bridge/PayloadCodec.cs
using System.Buffers.Binary;
using static Forge.Host.Bridge.PayloadFormat;

namespace Forge.Host.Bridge;

public static partial class BridgePayloadCodec
{
    public const uint MaxEchoDelayMs = 60_000;

    public static byte[] EncodeHandshake(HostHandshake handshake)
    {
        var writer = new PayloadWriter();
        writer.WriteString(handshake.HostVersion);
        writer.WriteString(handshake.SdkRevision);
        writer.WriteStrings(handshake.SupportedGames);
        writer.WriteStrings(handshake.Capabilities);
        return writer.ToArray();
    }

    public static HostHandshake DecodeHandshake(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new HostHandshake(
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadStrings(),
            reader.ReadStrings());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeEchoRequest(string message, uint delayMs = 0)
    {
        if (delayMs > MaxEchoDelayMs) Malformed("Echo delay exceeds limit");
        var writer = new PayloadWriter();
        writer.WriteUInt32(delayMs);
        writer.WriteString(message);
        return writer.ToArray();
    }

    public static EchoRequest DecodeEchoRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var delayMs = reader.ReadUInt32();
        if (delayMs > MaxEchoDelayMs) Malformed("Echo delay exceeds limit");
        var value = new EchoRequest(reader.ReadString(), delayMs);
        reader.Complete();
        return value;
    }

    public static byte[] EncodeText(string value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value);
        return writer.ToArray();
    }

    public static string DecodeText(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = reader.ReadString();
        reader.Complete();
        return value;
    }

    public static byte[] EncodeProgress(uint completed, uint total)
    {
        if (total == 0 || completed > total) Malformed("Invalid progress values");
        var payload = GC.AllocateUninitializedArray<byte>(8);
        BinaryPrimitives.WriteUInt32LittleEndian(payload, completed);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(4), total);
        return payload;
    }

    public static BridgeProgress DecodeProgress(ReadOnlySpan<byte> payload)
    {
        if (payload.Length != 8) Malformed("Progress payload must contain eight bytes");
        var completed = BinaryPrimitives.ReadUInt32LittleEndian(payload);
        var total = BinaryPrimitives.ReadUInt32LittleEndian(payload[4..]);
        if (total == 0 || completed > total) Malformed("Invalid progress values");
        return new(completed, total);
    }

    public static byte[] EncodeUyaIsoValidation(UyaIsoValidationPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteBoolean(value.IsSupported);
        writer.WriteString(value.Game);
        writer.WriteString(value.Region);
        writer.WriteString(value.Revision);
        writer.WriteString(value.Serial);
        writer.WriteUInt64(value.Size);
        writer.WriteString(value.Fingerprint);
        writer.WriteString(value.Diagnostic);
        return writer.ToArray();
    }

    public static UyaIsoValidationPayload DecodeUyaIsoValidation(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaIsoValidationPayload(
            reader.ReadBoolean(), reader.ReadString(), reader.ReadString(), reader.ReadString(),
            reader.ReadString(), reader.ReadUInt64(), reader.ReadString(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeDevelopmentIsoRequest(DevelopmentIsoRequest value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.SourcePath);
        writer.WriteString(value.TargetPath);
        writer.WriteString(value.Fingerprint);
        writer.WriteBoolean(value.Overwrite);
        return writer.ToArray();
    }

    public static DevelopmentIsoRequest DecodeDevelopmentIsoRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new DevelopmentIsoRequest(
            reader.ReadString(), reader.ReadString(), reader.ReadString(), reader.ReadBoolean());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeDevelopmentIso(DevelopmentIsoPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.Path);
        writer.WriteUInt64(value.Size);
        writer.WriteString(value.Fingerprint);
        return writer.ToArray();
    }

    public static DevelopmentIsoPayload DecodeDevelopmentIso(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new DevelopmentIsoPayload(reader.ReadString(), reader.ReadUInt64(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeUyaAssetImportRequest(UyaAssetImportRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.SourceIsoPath);
        writer.WriteString(value.CatalogRootPath);
        writer.WriteString(value.Fingerprint);
        writer.WriteString(value.Revision);
        writer.WriteBoolean(value.Force);
        return writer.ToArray();
    }

    public static UyaAssetImportRequestPayload DecodeUyaAssetImportRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaAssetImportRequestPayload(
            reader.ReadString(), reader.ReadString(), reader.ReadString(), reader.ReadString(), reader.ReadBoolean());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeUyaAssetImportResult(UyaAssetImportResultPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteUInt32(value.CompletedLevels);
        writer.WriteUInt32(value.TotalLevels);
        writer.WriteUInt32(value.AssetAppearances);
        writer.WriteUInt32(value.UniqueAssets);
        writer.WriteUInt32(value.FailedAssets);
        writer.WriteBoolean(value.Resumed);
        return writer.ToArray();
    }

    public static UyaAssetImportResultPayload DecodeUyaAssetImportResult(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaAssetImportResultPayload(
            reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadBoolean());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeUyaProjectOptions(UyaProjectOptionsPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteUInt32s(value.Levels);
        writer.WriteStrings(value.Warnings);
        return writer.ToArray();
    }

    public static UyaProjectOptionsPayload DecodeUyaProjectOptions(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaProjectOptionsPayload(reader.ReadUInt32s(), reader.ReadStrings());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeUyaProjectCreationRequest(UyaProjectCreationRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.SourceIsoPath);
        writer.WriteString(value.CatalogRootPath);
        writer.WriteString(value.ProjectPath);
        writer.WriteString(value.Name);
        writer.WriteString(value.Fingerprint);
        writer.WriteString(value.Revision);
        writer.WriteUInt32(value.Level);
        writer.WriteBoolean(value.AllowPartial);
        return writer.ToArray();
    }

    public static UyaProjectCreationRequestPayload DecodeUyaProjectCreationRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaProjectCreationRequestPayload(
            reader.ReadString(), reader.ReadString(), reader.ReadString(), reader.ReadString(),
            reader.ReadString(), reader.ReadString(), reader.ReadUInt32(), reader.ReadBoolean());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeUyaProjectPreflightRequest(UyaProjectPreflightRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.SourceIsoPath);
        writer.WriteString(value.CatalogRootPath);
        writer.WriteUInt32(value.Level);
        return writer.ToArray();
    }

    public static UyaProjectPreflightRequestPayload DecodeUyaProjectPreflightRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaProjectPreflightRequestPayload(reader.ReadString(), reader.ReadString(), reader.ReadUInt32());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeUyaProjectPreflight(UyaProjectPreflightPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteUInt32(value.Level);
        writer.WriteUInt32(value.SourceInstanceCount);
        writer.WriteUInt32(value.RenderableInstanceCount);
        writer.WriteUInt32(value.ModelLessInstanceCount);
        writer.WriteUInt32(value.MissingAssetInstanceCount);
        writer.WriteUInt32(value.MissingClassCount);
        writer.WriteStrings(value.Warnings);
        return writer.ToArray();
    }

    public static UyaProjectPreflightPayload DecodeUyaProjectPreflight(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaProjectPreflightPayload(
            reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(),
            reader.ReadUInt32(), reader.ReadStrings());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeProjectInspectRequest(ProjectInspectRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.ProjectPath);
        writer.WriteString(value.CatalogRootPath);
        return writer.ToArray();
    }

    public static ProjectInspectRequestPayload DecodeProjectInspectRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new ProjectInspectRequestPayload(reader.ReadString(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeProjectRenameRequest(ProjectRenameRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.ProjectPath);
        writer.WriteString(value.CatalogRootPath);
        writer.WriteString(value.Name);
        return writer.ToArray();
    }

    public static ProjectRenameRequestPayload DecodeProjectRenameRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new ProjectRenameRequestPayload(reader.ReadString(), reader.ReadString(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeProjectRecoveryRequest(ProjectRecoveryRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.ProjectPath);
        writer.WriteString(value.CatalogRootPath);
        writer.WriteString(value.RecoveryId);
        return writer.ToArray();
    }

    public static ProjectRecoveryRequestPayload DecodeProjectRecoveryRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new ProjectRecoveryRequestPayload(reader.ReadString(), reader.ReadString(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeForgeProjectDescriptor(ForgeProjectDescriptorPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.Path);
        writer.WriteString(value.Name);
        writer.WriteString(value.TargetGame);
        writer.WriteString(value.TargetRegion);
        writer.WriteString(value.TargetRevision);
        writer.WriteString(value.BakeProfile);
        writer.WriteUInt32(value.BaseLevel);
        writer.WriteUInt64(value.ModifiedUnixMilliseconds);
        writer.WriteUInt32(value.EntityCount);
        writer.WriteUInt32(value.MissingAssetCount);
        writer.WriteBoolean(value.IsDirty);
        writer.WriteBoolean(value.MigrationPending);
        writer.WriteStrings(value.Warnings);
        if (value.Recoveries.Count > MaxListItems) Malformed("List exceeds item limit");
        writer.WriteUInt32((uint)value.Recoveries.Count);
        foreach (var recovery in value.Recoveries)
        {
            writer.WriteString(recovery.Id);
            writer.WriteUInt64(recovery.CreatedUnixMilliseconds);
            writer.WriteString(recovery.Name);
            writer.WriteUInt32(recovery.EntityCount);
            writer.WriteString(recovery.Fingerprint);
            writer.WriteUInt64(recovery.Size);
        }
        if (value.MissingAssets.Count > 4_096) Malformed("Missing-asset list exceeds item limit");
        writer.WriteUInt32((uint)value.MissingAssets.Count);
        foreach (var missing in value.MissingAssets)
        {
            writer.WriteString(missing.Id);
            writer.WriteString(missing.Kind);
            writer.WriteUInt32(missing.EntityCount);
            writer.WriteBoolean(missing.Repairable);
            writer.WriteStrings(missing.Provenance);
        }
        return writer.ToArray();
    }

    public static ForgeProjectDescriptorPayload DecodeForgeProjectDescriptor(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var path = reader.ReadString();
        var name = reader.ReadString();
        var targetGame = reader.ReadString();
        var targetRegion = reader.ReadString();
        var targetRevision = reader.ReadString();
        var bakeProfile = reader.ReadString();
        var baseLevel = reader.ReadUInt32();
        var modified = reader.ReadUInt64();
        var entityCount = reader.ReadUInt32();
        var missingAssetCount = reader.ReadUInt32();
        var isDirty = reader.ReadBoolean();
        var migrationPending = reader.ReadBoolean();
        var warnings = reader.ReadStrings();
        var recoveryCount = reader.ReadUInt32();
        if (recoveryCount > MaxListItems) Malformed("List exceeds item limit");
        var recoveries = new ProjectRecoverySnapshotPayload[recoveryCount];
        for (var index = 0; index < recoveries.Length; index++)
        {
            recoveries[index] = new(
                reader.ReadString(), reader.ReadUInt64(), reader.ReadString(), reader.ReadUInt32(), reader.ReadString(), reader.ReadUInt64());
        }
        var missingCount = reader.ReadUInt32();
        if (missingCount > 4_096) Malformed("Missing-asset list exceeds item limit");
        var missingAssets = new MissingProjectAssetPayload[missingCount];
        for (var index = 0; index < missingAssets.Length; index++)
        {
            missingAssets[index] = new(
                reader.ReadString(), reader.ReadString(), reader.ReadUInt32(), reader.ReadBoolean(), reader.ReadStrings());
        }
        var value = new ForgeProjectDescriptorPayload(
            path, name, targetGame, targetRegion, targetRevision, bakeProfile, baseLevel, modified,
            entityCount, missingAssetCount, isDirty, migrationPending, warnings, recoveries, missingAssets);
        reader.Complete();
        return value;
    }

}

File: src/main/bridge/PayloadCodec.ts
import { Buffer } from 'node:buffer';

import type {
  DevelopmentIsoResult,
  ForgeHostStatus,
  ForgeProjectDescriptor,
  Progress,
  UyaAssetImportResult,
  UyaIsoIdentity,
  UyaProjectOptions,
  UyaProjectPreflight,
} from '../../types/ForgeApi.js';
import { MAX_LIST_ITEMS, malformed, PayloadReader, PayloadWriter } from './PayloadIO.ts';
import type {
  DevelopmentIsoRequest,
  EchoRequest,
  ProjectInspectRequest,
  ProjectRecoveryRequest,
  ProjectRenameRequest,
  UyaAssetImportRequest,
  UyaProjectCreationRequest,
  UyaProjectPreflightRequest,
} from '../../types/BridgePayloads.js';

export const MAX_ECHO_DELAY_MS = 60_000;

export function encodeHandshake(value: ForgeHostStatus): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.hostVersion);
  writer.writeString(value.sdkRevision);
  writer.writeStrings(value.supportedGames);
  writer.writeStrings(value.capabilities);
  return writer.toBuffer();
}

export function decodeHandshake(payload: Uint8Array): ForgeHostStatus {
  const reader = new PayloadReader(payload);
  const value = {
    hostVersion: reader.readString(),
    sdkRevision: reader.readString(),
    supportedGames: reader.readStrings(),
    capabilities: reader.readStrings(),
  };
  reader.complete();
  return value;
}

export function encodeEchoRequest(message: string, delayMs = 0): Buffer {
  if (!Number.isInteger(delayMs) || delayMs < 0 || delayMs > MAX_ECHO_DELAY_MS) malformed('Invalid echo delay');
  const writer = new PayloadWriter();
  writer.writeUInt32(delayMs);
  writer.writeString(message);
  return writer.toBuffer();
}

export function decodeEchoRequest(payload: Uint8Array): EchoRequest {
  const reader = new PayloadReader(payload);
  const delayMs = reader.readUInt32();
  if (delayMs > MAX_ECHO_DELAY_MS) malformed('Echo delay exceeds limit');
  const message = reader.readString();
  reader.complete();
  return { message, delayMs };
}

export function encodeText(value: string): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value);
  return writer.toBuffer();
}

export function decodeText(payload: Uint8Array): string {
  const reader = new PayloadReader(payload);
  const value = reader.readString();
  reader.complete();
  return value;
}

export function encodeProgress(completed: number, total: number): Buffer {
  if (!Number.isInteger(completed) || !Number.isInteger(total) || total <= 0 || completed < 0 || completed > total) {
    malformed('Invalid progress values');
  }
  const payload = Buffer.allocUnsafe(8);
  payload.writeUInt32LE(completed, 0);
  payload.writeUInt32LE(total, 4);
  return payload;
}

export function decodeProgress(payload: Uint8Array): Progress {
  if (payload.length !== 8) malformed('Progress payload must contain eight bytes');
  const bytes = Buffer.from(payload.buffer, payload.byteOffset, payload.byteLength);
  const completed = bytes.readUInt32LE(0);
  const total = bytes.readUInt32LE(4);
  if (total === 0 || completed > total) malformed('Invalid progress values');
  return { completed, total };
}

export function encodeUyaIsoValidation(value: UyaIsoIdentity): Buffer {
  const writer = new PayloadWriter();
  writer.writeBoolean(value.isSupported);
  writer.writeString(value.game);
  writer.writeString(value.region);
  writer.writeString(value.revision);
  writer.writeString(value.serial);
  writer.writeUInt64(value.size);
  writer.writeString(value.fingerprint);
  writer.writeString(value.diagnostic);
  return writer.toBuffer();
}

export function decodeUyaIsoValidation(payload: Uint8Array): UyaIsoIdentity {
  const reader = new PayloadReader(payload);
  const value = {
    isSupported: reader.readBoolean(),
    game: reader.readString(),
    region: reader.readString(),
    revision: reader.readString(),
    serial: reader.readString(),
    size: reader.readUInt64(),
    fingerprint: reader.readString(),
    diagnostic: reader.readString(),
  };
  reader.complete();
  return value;
}

export function encodeDevelopmentIsoRequest(value: DevelopmentIsoRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.sourcePath);
  writer.writeString(value.targetPath);
  writer.writeString(value.fingerprint);
  writer.writeBoolean(value.overwrite);
  return writer.toBuffer();
}

export function decodeDevelopmentIsoRequest(payload: Uint8Array): DevelopmentIsoRequest {
  const reader = new PayloadReader(payload);
  const value = {
    sourcePath: reader.readString(),
    targetPath: reader.readString(),
    fingerprint: reader.readString(),
    overwrite: reader.readBoolean(),
  };
  reader.complete();
  return value;
}

export function encodeDevelopmentIso(value: DevelopmentIsoResult): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.path);
  writer.writeUInt64(value.size);
  writer.writeString(value.fingerprint);
  return writer.toBuffer();
}

export function decodeDevelopmentIso(payload: Uint8Array): DevelopmentIsoResult {
  const reader = new PayloadReader(payload);
  const value = { path: reader.readString(), size: reader.readUInt64(), fingerprint: reader.readString() };
  reader.complete();
  return value;
}

export function encodeUyaAssetImportRequest(value: UyaAssetImportRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.sourceIsoPath);
  writer.writeString(value.catalogRootPath);
  writer.writeString(value.fingerprint);
  writer.writeString(value.revision);
  writer.writeBoolean(value.force);
  return writer.toBuffer();
}

export function decodeUyaAssetImportRequest(payload: Uint8Array): UyaAssetImportRequest {
  const reader = new PayloadReader(payload);
  const value = {
    sourceIsoPath: reader.readString(),
    catalogRootPath: reader.readString(),
    fingerprint: reader.readString(),
    revision: reader.readString(),
    force: reader.readBoolean(),
  };
  reader.complete();
  return value;
}

export function encodeUyaAssetImportResult(value: UyaAssetImportResult): Buffer {
  const writer = new PayloadWriter();
  writer.writeUInt32(value.completedLevels);
  writer.writeUInt32(value.totalLevels);
  writer.writeUInt32(value.assetAppearances);
  writer.writeUInt32(value.uniqueAssets);
  writer.writeUInt32(value.failedAssets);
  writer.writeBoolean(value.resumed);
  return writer.toBuffer();
}

export function decodeUyaAssetImportResult(payload: Uint8Array): UyaAssetImportResult {
  const reader = new PayloadReader(payload);
  const value = {
    completedLevels: reader.readUInt32(),
    totalLevels: reader.readUInt32(),
    assetAppearances: reader.readUInt32(),
    uniqueAssets: reader.readUInt32(),
    failedAssets: reader.readUInt32(),
    resumed: reader.readBoolean(),
  };
  reader.complete();
  return value;
}

export function encodeUyaProjectOptions(value: UyaProjectOptions): Buffer {
  const writer = new PayloadWriter();
  writer.writeUInt32s(value.levels);
  writer.writeStrings(value.warnings);
  return writer.toBuffer();
}

export function decodeUyaProjectOptions(payload: Uint8Array): UyaProjectOptions {
  const reader = new PayloadReader(payload);
  const value = { levels: reader.readUInt32s(), warnings: reader.readStrings() };
  reader.complete();
  return value;
}

export function encodeUyaProjectCreationRequest(value: UyaProjectCreationRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.sourceIsoPath);
  writer.writeString(value.catalogRootPath);
  writer.writeString(value.projectPath);
  writer.writeString(value.name);
  writer.writeString(value.fingerprint);
  writer.writeString(value.revision);
  writer.writeUInt32(value.level);
  writer.writeBoolean(value.allowPartial);
  return writer.toBuffer();
}

export function decodeUyaProjectCreationRequest(payload: Uint8Array): UyaProjectCreationRequest {
  const reader = new PayloadReader(payload);
  const value = {
    sourceIsoPath: reader.readString(),
    catalogRootPath: reader.readString(),
    projectPath: reader.readString(),
    name: reader.readString(),
    fingerprint: reader.readString(),
    revision: reader.readString(),
    level: reader.readUInt32(),
    allowPartial: reader.readBoolean(),
  };
  reader.complete();
  return value;
}

export function encodeUyaProjectPreflightRequest(value: UyaProjectPreflightRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.sourceIsoPath);
  writer.writeString(value.catalogRootPath);
  writer.writeUInt32(value.level);
  return writer.toBuffer();
}

export function decodeUyaProjectPreflightRequest(payload: Uint8Array): UyaProjectPreflightRequest {
  const reader = new PayloadReader(payload);
  const value = { sourceIsoPath: reader.readString(), catalogRootPath: reader.readString(), level: reader.readUInt32() };
  reader.complete();
  return value;
}

export function encodeUyaProjectPreflight(value: UyaProjectPreflight): Buffer {
  const writer = new PayloadWriter();
  writer.writeUInt32(value.level);
  writer.writeUInt32(value.sourceInstanceCount);
  writer.writeUInt32(value.renderableInstanceCount);
  writer.writeUInt32(value.modelLessInstanceCount);
  writer.writeUInt32(value.missingAssetInstanceCount);
  writer.writeUInt32(value.missingClassCount);
  writer.writeStrings(value.warnings);
  return writer.toBuffer();
}

export function decodeUyaProjectPreflight(payload: Uint8Array): UyaProjectPreflight {
  const reader = new PayloadReader(payload);
  const value = {
    level: reader.readUInt32(),
    sourceInstanceCount: reader.readUInt32(),
    renderableInstanceCount: reader.readUInt32(),
    modelLessInstanceCount: reader.readUInt32(),
    missingAssetInstanceCount: reader.readUInt32(),
    missingClassCount: reader.readUInt32(),
    warnings: reader.readStrings(),
  };
  reader.complete();
  return value;
}

export function encodeProjectInspectRequest(value: ProjectInspectRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.projectPath);
  writer.writeString(value.catalogRootPath);
  return writer.toBuffer();
}

export function decodeProjectInspectRequest(payload: Uint8Array): ProjectInspectRequest {
  const reader = new PayloadReader(payload);
  const value = { projectPath: reader.readString(), catalogRootPath: reader.readString() };
  reader.complete();
  return value;
}

export function encodeProjectRenameRequest(value: ProjectRenameRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.projectPath);
  writer.writeString(value.catalogRootPath);
  writer.writeString(value.name);
  return writer.toBuffer();
}

export function decodeProjectRenameRequest(payload: Uint8Array): ProjectRenameRequest {
  const reader = new PayloadReader(payload);
  const value = { projectPath: reader.readString(), catalogRootPath: reader.readString(), name: reader.readString() };
  reader.complete();
  return value;
}

export function encodeProjectRecoveryRequest(value: ProjectRecoveryRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.projectPath);
  writer.writeString(value.catalogRootPath);
  writer.writeString(value.recoveryId);
  return writer.toBuffer();
}

export function decodeProjectRecoveryRequest(payload: Uint8Array): ProjectRecoveryRequest {
  const reader = new PayloadReader(payload);
  const value = { projectPath: reader.readString(), catalogRootPath: reader.readString(), recoveryId: reader.readString() };
  reader.complete();
  return value;
}

export function encodeForgeProjectDescriptor(value: ForgeProjectDescriptor): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.path);
  writer.writeString(value.name);
  writer.writeString(value.targetGame);
  writer.writeString(value.targetRegion);
  writer.writeString(value.targetRevision);
  writer.writeString(value.bakeProfile);
  writer.writeUInt32(value.baseLevel);
  writer.writeUInt64(value.modifiedUnixMilliseconds);
  writer.writeUInt32(value.entityCount);
  writer.writeUInt32(value.missingAssetCount);
  writer.writeBoolean(value.isDirty);
  writer.writeBoolean(value.migrationPending);
  writer.writeStrings(value.warnings);
  if (value.recoveries.length > MAX_LIST_ITEMS) malformed('List exceeds item limit');
  writer.writeUInt32(value.recoveries.length);
  value.recoveries.forEach((recovery) => {
    writer.writeString(recovery.id);
    writer.writeUInt64(recovery.createdUnixMilliseconds);
    writer.writeString(recovery.name);
    writer.writeUInt32(recovery.entityCount);
    writer.writeString(recovery.fingerprint);
    writer.writeUInt64(recovery.size);
  });
  if (value.missingAssets.length > 4_096) malformed('Missing-asset list exceeds item limit');
  writer.writeUInt32(value.missingAssets.length);
  value.missingAssets.forEach((missing) => {
    writer.writeString(missing.id);
    writer.writeString(missing.kind);
    writer.writeUInt32(missing.entityCount);
    writer.writeBoolean(missing.repairable);
    writer.writeStrings(missing.provenance);
  });
  return writer.toBuffer();
}

export function decodeForgeProjectDescriptor(payload: Uint8Array): ForgeProjectDescriptor {
  const reader = new PayloadReader(payload);
  const value: ForgeProjectDescriptor = {
    path: reader.readString(),
    name: reader.readString(),
    targetGame: reader.readString(),
    targetRegion: reader.readString(),
    targetRevision: reader.readString(),
    bakeProfile: reader.readString(),
    baseLevel: reader.readUInt32(),
    modifiedUnixMilliseconds: reader.readUInt64(),
    entityCount: reader.readUInt32(),
    missingAssetCount: reader.readUInt32(),
    isDirty: reader.readBoolean(),
    migrationPending: reader.readBoolean(),
    warnings: reader.readStrings(),
    recoveries: [],
    missingAssets: [],
  };
  const recoveryCount = reader.readUInt32();
  if (recoveryCount > MAX_LIST_ITEMS) malformed('List exceeds item limit');
  value.recoveries = Array.from({ length: recoveryCount }, () => ({
    id: reader.readString(),
    createdUnixMilliseconds: reader.readUInt64(),
    name: reader.readString(),
    entityCount: reader.readUInt32(),
    fingerprint: reader.readString(),
    size: reader.readUInt64(),
  }));
  const missingCount = reader.readUInt32();
  if (missingCount > 4_096) malformed('Missing-asset list exceeds item limit');
  value.missingAssets = Array.from({ length: missingCount }, () => ({
    id: reader.readString(),
    kind: reader.readString(),
    entityCount: reader.readUInt32(),
    repairable: reader.readBoolean(),
    provenance: reader.readStrings(),
  }));
  reader.complete();
  return value;
}
